### ISSUE TITLE
Revisit Annotation popover implementation to better support our current annotation work

### ISSUE THREAD

**Issue by amoeba** [NCEAS/metacatui#1807]
Most of our work with semantic annotations has been focused around adding `oboe:containsMeasurementsOfType` annotations so the UI is designed around that use case. We've always had other types of annotations but we'll be using them more and I thought it'd be good to re-visit the design to see if it could be improved. For example, @laijasmine and the team are issuing annotations with new properties and objects. See the current UI with one of these in-draft annotations:

<img width="401" alt="Screen Shot 2021-05-25 at 11 58 40 AM" src="https://user-images.githubusercontent.com/563/119560822-e57add80-bd50-11eb-8621-dcfa3a155e4e.png">

This is pretty useful IMO but you don't get any information about the property "madeBySystem" and you have to be pretty lucky to find the link embedded in the popover to de-reference it. When properties have good, human-readable labels, this isn't a huge deal but (1) labels aren't always going to be enough and (2) we're stopping short of really exposing all the power of the property by helping the user de-reference it.

I think a next-gen interface could:

- Try to grab a definition for the property from one or more external sources and show it in-line like we already do for the object of the triple
- Not show the "No extra information..." statement and just show the "Find more information..." statement.
- Show a better visual indicator of loading progress

**Comment by mpsaloha** [comment]
Agreed! -  that we should enable some better understanding of the
predicates, which can be at least somewhat addressed by adding Annotations
(dc:description; rdfs:comment) to the Object and Datatype Properties that
are currently defined in the MOSAiC ontology.  (but I was too busy building
the Ontology to comprehensively do this on a first pass)

Ironically, for the example shown with predicate "*madeBySystem*" I had
created an EquivalentProperty to *SSN:deployedSystem*, where there are
formal definitions of what the predicate means, but these are probably too
complicated to be useful for our stakeholders.

Also, again using the above example, there is MUCH MORE information in the
current draft MOSAiC Ontology-- e.g. that the "Accoustical Backscatter
Sensor (ABS)" (note unfortunate misspelling by MOSAiC of "Acoustical"; and
there were more!)-- has two Instances-- but taking one as an example-- a
'Scintec Sodar' device has a MOSAiC SensorID of '4242', and a descriptive
webpage at  'https://sensor.awi.de/?id=4242'.  All of this information is
in the Ontology!  That is, we could (I'm pretty sure, but Jasmine could
correct me) clarify the exact sensor device that created some set of
measurements in an Event, along with a description of that device.  All the
requisite information linking the Event to the specific device used, and
its documentation page, are already in the Ontology.

Bottom line-- we should try to add some *dc:description* or *rdfs:comment*
to all our Classes and Properties.  But we should also consider how much of
the Ontology contents we want to make accessible to our stakeholders, as
the MOSAiC ontology is quite rich, and could serve much like an online
Metadata Catalogue.

cheers,
Mark

**Comment by amoeba** [comment]
Thanks @mpsaloha, your comment about filling in proper comments/labels for all terms is a big deal. Both generally but also specifically for our current implementation of this UI widget. Having as near to 100% coverage for labels on all terms makes the whole thing work much better.

Re: The changes I am wanting to make, I've had time to play with some ideas and I'm leaning towards simply having one popover for the property and another for the label.

_Note: These are mockups are halfway between wireframes and full mockups so the colors and much of the style would change when implemented._

<img width="781" alt="Screen Shot 2021-06-02 at 4 20 07 PM" src="https://user-images.githubusercontent.com/563/120567574-9e2ec580-c3be-11eb-895f-151d6e24bd3b.png">

In addition, the new popover would have a standard loading indicator. Right now, we show static text "Looking for more information..." but I don't think it's all that clear anything is happening.

So the other tweak we could make would be to put a spinner or flashing text in here:

<img width="423" alt="Screen Shot 2021-06-02 at 4 33 24 PM" src="https://user-images.githubusercontent.com/563/120568455-5c068380-c3c0-11eb-9bab-aeb3f5824c3f.png">

My work is at https://www.figma.com/file/U9TqRVCDpGTG59xrHJ5Kto/Annotation-Popover-Re-Design?node-id=0%3A1 if anyone wants to take a look. **The main question I have at this point is whether it makes more sense to have a combine mega-popover or two separate ones**. I'm leaning separate.

**Comment by mbjones** [comment]
This looks good @amoeba. I like the separate popups in `1b` better - I think they are clearer and more directly linked to the UI to make them understandable.

In your figma mockup, I see that when the term lookup fails to find more information, you display a string like `Find more information about **madeBySystem** at MOSAIC:00002250`. I think, rather than having a loading indicator while you are doing the lookup, you could start with that default text, and only replace it with the definitions if something is found on lookup. I find loading icons distracting, and its better to provide as much info as possible as early as possible. What do you think?

**Comment by amoeba** [comment]
Now that you say it, I think that's a great idea. Having a loader take over an entire section of the popover does have a few downsides. I think I'll go with the route you suggest when we do this.

**Comment by amoeba** [comment]
Our MOSAiC work has brought up another use case. For ECSO terms, we mainly wanted to provide a definition and a link for the user to get more info. With MOSAiC, we're using a lot of instances so they don't have definitions but instead have lots of interesting object and datatype properties we could show. See how things like in Protégé:

<img width="637" alt="Screen Shot 2021-06-25 at 3 51 32 PM" src="https://user-images.githubusercontent.com/563/123494705-435b4900-d5cd-11eb-9a35-a044d6a5e04d.png">

Some things to look into;

- Can we get some of this information with BioPortal's API?
- How about OLS or COR?
- Should we think about hosting our own triplestore so we have more control over what information we can get? Are the other solutions?

**Comment by mbjones** [comment]
We plan to host our own triple store with slinky.... so we will likely have a copy of all relevant ontologies loaded there. Will that help?

**Comment by amoeba** [comment]
As of https://github.com/NCEAS/metacatui/commit/cc2324ac877a416316c1a749b217348fd947e49c, both parts of the annotation pill have popovers, see:

<img width="484" alt="Screen Shot 2021-07-28 at 4 21 34 PM" src="https://user-images.githubusercontent.com/563/127412672-a71acb66-1f6c-4bd6-bc1d-e158c8e7d968.png">

This matches the direction we went with the discussion above.

With the corresponding changes in https://github.com/DataONEorg/d1_cn_index_processor/issues/27 and https://github.com/NCEAS/metacat/issues/1517, the popover for the property can now find other datasets that annotate with that property (and any superproperties). Similar to what we do with values now.

@laurenwalker do you want a chance to review this? I'm happy to PR if so or merge otherwise.

**Comment by laurenwalker** [comment]
Thanks Bryce. Do your code changes depend on a Metacat upgrade for [NCEAS/metacat#1517](https://github.com/NCEAS/metacat/issues/1517)?

**Comment by laurenwalker** [comment]
Nevermind, I see in your commit message that they are dependent on a Metacat upgrade. Will the AnnotationView have a fallback display for Metacat 2.15.0 and earlier? I think it would be good to make sure the annotations and their popovers aren't completely broken if someone were to upgrade MetacatUI but not Metacat. The fallback could be temporary just to give people a chance to upgrade Metacat.

**Comment by mbjones** [comment]
I think requiring the metacat dependency is fine, and probably unavoidable as the indexing changes are needed to power the feature. So, at most I think the fallback could be that the feature is disabled for older versions of Metacat (maybe the old behaviour could be maintained if its not too complicated to keep). 

In general, though, we've tried to not maintain older versions because of the technical debt it introduces into the codebase of conditioning behavior on installation versions. We try to have a single version at the HEAD that we are maintaining. So, having a hard requirement on at least Metacat 2.16 or whatever it ends up as is probably ok.

**Comment by laurenwalker** [comment]
Ok, we can require a Metacat upgrade for this, but I do want to make sure at least the annotation view fails gracefully or has some kind of basic fallback display if the Metacat is older (e.g. displaying the annotation text without a popover to get the definition, ontology link, etc.) Bryce may have already thought of this - I haven't checked out the feature branch yet.

**Comment by laurenwalker** [comment]
Because it requires a Metacat upgrade, @amoeba could you submit a PR and I'll merge once Metacat 2.16.0 is ready for release?

**Comment by amoeba** [comment]
I did think about fallback behavior here so the change in https://github.com/NCEAS/metacat/issues/1517 was made to be backwards-compatible with older versions of MetacatUI. On older versions, you get the current behavior (one popover that focuses on the `valueURI`). On versions with this changeset, you get the new behavior (separate popovers). 

One combo I didn't cover is when you have an older Metacat version without https://github.com/NCEAS/metacat/issues/1517 but _do_ have a newer MetacatUI w/ the updates in this changeset. It looks a little broken because I had to factor the underlying DOM structure coming from Metacat's view XSLT. I'd like to make it not broken and I think it'd be a small change.

@laurenwalker can I make one more change so the above thing works better and then I'll send in a PR and ping you?

**Comment by laurenwalker** [comment]
Thanks Bryce, sounds great! Thanks for checking

**Comment by amoeba** [comment]
I put in some basic fallback code to just disable the popover behavior when the new code encounters an old Metacat and that cleans up the DOM so it at least looks right. That rounds out the fourth version combination:

| | Old Metacat | New Metacat |
|-:|:-:|:-:|
|**Old MetacatUI** | ✅ | Pill has one minor glitch, old popover works |
|**New MetacatUI** | Pill looks good, no popovers | ✅ |

The minor glitch above when an old MetacatUI encounters an upgraded Metacat is just that the right side of the pill doesn't have a rounded border. This could be fixed but I wanted to stop here to avoid adding too much detection code.

PR forthcoming.

**Comment by amoeba** [comment]
PRd at https://github.com/NCEAS/metacatui/pull/1843 @laurenwalker and ready for review.
